src/lib/ast.rs

```diff
@@ -0,0 +1,623 @@
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::fs::File;
+use std::io;
+use std::io::{Read, Write};
+use std::ops::{Index, IndexMut};
+use std::path::Path;
+
+use dot;
+use lrlex::{build_lex, Lexeme};
+use lrpar::parser;
+use lrtable::{Grammar, Minimiser, TIdx, yacc_to_statetable};
+
+#[derive(Debug)]
+/// Errors raised when parsing a source file.
+pub enum ParseError {
+    /// Io error returned from standard library routine.
+    Io(io::Error),
+    /// File not found.
+    FileNotFound,
+    /// File name had no extension (used to determine which lexer/parser to use).
+    NoFileExtension,
+    /// Lexer could not be built by `lrlex`.
+    BrokenLexer,
+    /// Parser could not be built by `lrpar`.
+    BrokenParser,
+    /// File contained lexical error and could not be lexed.
+    LexicalError,
+    /// File contained syntax error and could not be parsed.
+    SyntaxError,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd)]
+/// A node identifier for nodes within a particular `Arena`.
+pub struct NodeId {
+    index: usize,
+}
+
+/// A type to represent edges between nodes within an `Arena`.
+type EdgeId = (usize, usize);
+
+impl fmt::Display for NodeId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:}", self.index)
+    }
+}
+
+/// An arena in which to place `NodeID`-indexed AST nodes.
+pub struct Arena<T, U> {
+    nodes: Vec<Node<T, U>>,
+}
+
+impl<T, U> Index<NodeId> for Arena<T, U> {
+    type Output = Node<T, U>;
+
+    fn index(&self, node: NodeId) -> &Node<T, U> {
+        &self.nodes[node.index]
+    }
+}
+
+impl<T, U> IndexMut<NodeId> for Arena<T, U> {
+    fn index_mut(&mut self, node: NodeId) -> &mut Node<T, U> {
+        &mut self.nodes[node.index]
+    }
+}
+
+impl<T, U> Arena<T, U> {
+    /// Create an empty `Arena`.
+    pub fn new() -> Arena<T, U> {
+        Arena { nodes: Vec::new() }
+    }
+
+    /// Create a new node from its data.
+    pub fn new_node(&mut self, data: T, ty: U, indent: u16) -> NodeId {
+        let next_index = self.nodes.len();
```

Comment:
Here's the disassembly for `len` on a vector:
```
[0x00000b40 0% 5019 (0x0:-1=1)]> pd $r @ sym.__LT_collections..vec..Vec_LT_T_GT__GT_::len::hda194d93e28bba2b                          
/ (fcn) sym.__LT_collections..vec..Vec_LT_T_GT__GT_::len::hda194d93e28bba2b 26                                                        
|   sym.__LT_collections..vec..Vec_LT_T_GT__GT_::len::hda194d93e28bba2b ();                                                           
|           ; var int local_10h @ rbp-0x10                                                                                            
|           ; var int local_8h @ rbp-0x8                                                                                              
|           ; CALL XREF from 0x00001e5c (sym.hello::print_len::h1bd4f4b75b398548)                                                     
|           0x00000b40   *  55             push rbp                                                                                   
|           0x00000b41      4889e5         mov rbp, rsp                                                                               
|           0x00000b44      48897df0       mov qword [rbp - local_10h], rdi                                                           
|           0x00000b48      488b7df0       mov rdi, qword [rbp - local_10h]                                                           
|           0x00000b4c      48897df8       mov qword [rbp - local_8h], rdi                                                            
|           0x00000b50      488b7df8       mov rdi, qword [rbp - local_8h]                                                            
|           0x00000b54      488b4710       mov rax, qword [rdi + 0x10] ; [0x10:8]=0x1003e0003                                         
|           0x00000b58      5d             pop rbp                                                                                    
\           0x00000b59      c3             ret                                                                                        
            0x00000b5a      90             nop                                                                                        
            0x00000b5b      90             nop                                                                                        
            0x00000b5c      90             nop                                                                                        
            0x00000b5d      90             nop                                                                                        
            0x00000b5e      90             nop                                                                                        
            0x00000b5f      90             nop                                                                                        ```

There's no loop, so there is probably a count in the vec, or the compiler figured out the invariant.

Test program I extracted that code from:
```
extern crate rand;

use rand::Rng;

fn print_len(v: Vec<u16>) {
    println!("size={}", v.len());
}

fn main() {
    let mut rng = rand::thread_rng();
    let sz = rng.gen::<u16>();
    let mut v = Vec::new();
    for i in 0..sz {
        v.push(i);
    }
    print_len(v);

}
```